packages/flutter/lib/src/widgets/router.dart

```diff
@@ -0,0 +1,1257 @@
+///
+/// This provider is responsible for handing the route information through [value]
+/// getter and notifies listeners, typically the [Router] widget, when a new
+/// route information is available.
+///
+/// When the router opts for the route information reporting (by overrides the
+/// [RouterDelegate.currentConfiguration] to return non-null), overrides the
+/// [routerReportsNewRouteInformation] method to process the route information.
+///
+/// See also:
+///
+///  * [PlatformRouteInformationProvider], which wires up the itself with the
+///    [WidgetsBindingObserver.didPushRoute] to propagate platform push route
+///    intent to the [Router] widget, as well as reports new route information
+///    from the [Router] back to the engine by overriding the
+///    [routerReportsNewRouteInformation].
+abstract class RouteInformationProvider extends ValueListenable<RouteInformation> {
+  /// A callback called when the [Router] widget detects any navigation event
+  /// due to state changes.
+  ///
+  /// The subclasses can override this method to update theirs values or trigger
+  /// other side effects. For example, the [PlatformRouteInformationProvider]
+  /// overrides this method to report the route information back to the engine.
+  ///
+  /// The [routeInformation] is the new route information after the navigation
+  /// event.
+  void routerReportsNewRouteInformation(RouteInformation routeInformation) {}
+}
+
+/// The route information provider that propagates the platform route information changes.
+///
+/// This provider also reports the new route information from the [Router] widget
+/// back to engine using message channel method, the
+/// [RouteNotificationMessages.maybeNotifyRouteChange].
+class PlatformRouteInformationProvider extends RouteInformationProvider with WidgetsBindingObserver, ChangeNotifier {
+  /// Create a platform route information provider.
+  ///
+  /// Use the [initialRouteInformation] to set the default route information for this
+  /// provider.
+  PlatformRouteInformationProvider({RouteInformation initialRouteInformation})
+      : _value = initialRouteInformation,
+        _currentRouteInformationInEngine = initialRouteInformation;
+
+  RouteInformation _currentRouteInformationInEngine;
+
+  @override
+  void routerReportsNewRouteInformation(RouteInformation routeInformation) {
+    RouteNotificationMessages.maybeNotifyRouteInformationChange(
+      routeInformation,
+      _currentRouteInformationInEngine
+    );
+    _currentRouteInformationInEngine = routeInformation;
+    _value = routeInformation;
+  }
+
+  void _updateAndNotify(RouteInformation newValue) {
+    assert(_value != newValue);
+    _value = newValue;
+    notifyListeners();
+  }
+
+  @override
+  RouteInformation get value => _value;
+  RouteInformation _value;
+  set value(RouteInformation newValue) {
+    if (_value == newValue)
+      return;
+    _updateAndNotify(newValue);
+    routerReportsNewRouteInformation(newValue);
+  }
```

Comment:
What's the setter for? Shouldn't you always go through routerReportsNewRouteInformation?

Reply:
This setter is for developers who want to navigate to a different location by changing the PlatformRouteInformationProvider.value directly.

In this case, we need to notify the router and report this event to the engine directly.

I think this is a very useful feature, because developers can still build up browser history entries even if they didn't implement the routerdelegate.currentConfiguration

Reply:
That needs to be documented. Also, should that then be called debugSetRouteInformation and only work in debug mode?

Reply:
Actually, I think we should just remove that. The feature goes against what Router wants to archive: That the current configuration of the navigator depends on the app state. If you want this, you should make your app state contain a routename string.

If your app really needs it, you can implement your own RouteInformationProvider with this functionality.